react/src/component/login/login.tsx

```diff
@@ -0,0 +1,217 @@
+            )
+        }
+        </>
+    )
+}
+
+export const NavbarAuthStatus = (props) => {
+    let loginStatus = useContext(CLoginStatus);
+    let schema = useContext(CSchema);
+    let appName = useContext(CAppName);
+    let appWrapper = new AppWrapper(appName);
+
+    if(loginStatus.method === "guest") {
+        return (<></>)
+    }
+
+    if(loginStatus.isLoggedIn) {
+        return (
+            <NavToggleDropdown title={
+                <>{
+                    loginStatus.avatarUrl ? (
+                        <Image src={loginStatus.avatarUrl} fluid rounded={true}/>
+                    ) : (
+                        <FontAwesomeIcon icon={Icon.faPerson}/>
+                    )
+                }</>
+            }>
+                {
+                    loginStatus.displayName && loginStatus.displayName.length > 0 && (
+                        <Dropdown.Header>{loginStatus.displayName}</Dropdown.Header>
+                    )
+                }
+                {
+                    loginStatus.paymentPlan && (
+                        <Dropdown.Header>{appWrapper.getPaymentPlanName(loginStatus.paymentPlan, schema)}</Dropdown.Header>
+                    )
+                }
+                {
+                    loginStatus.userName && (
+                        <Dropdown.Header>{loginStatus.userName}</Dropdown.Header>
+                    )
+                }
+                <Dropdown.Item href={`/${appName}/logout`}>Sign Out</Dropdown.Item>
+            </NavToggleDropdown>
+        )
+    }
+
+    return (
+        <Nav.Link href={`/${appName}/login`}>Sign In</Nav.Link>
+    )
+
+
+}
+
+export const CatchLoggedOut = (props) => {
+    let appName = useContext(CAppName);
+    let loginStatus = useContext(CLoginStatus);
+
+    return (
+        <>
+            {
+                (loginStatus.isLoggedIn && !loginStatus.needCompleteRegistration) || loginStatus.method === "guest" ?
+                (
+                    props.children
+                ) : (
+                    <Navigate to={`/react/${appName}/login`}/> // TODO @garsuga: abstract
+                )
+            }
+        </>
+    )
+}
+
+type LoginConfirmParams = {
+    token: string,
+    method: AuthMethod,
+    needCompleteRegistration: "0" | "1"
+}
+
+export const LoginConfirm = (props) => {
+    let { token, method, needCompleteRegistration } = useParams<LoginConfirmParams>();
+
+    switch(method) {
+        case 'email':
+            return <LoginEmailConfirm needCompleteRegistration={needCompleteRegistration} token={token}/>
+        default:
+            throw new Error(`could not handle login method=${method}`)
+    }
+}
+
+
+
+export const LoginNeedCompleteRegistration = (props) => {
+    let appName = useContext(CAppName);
+    let navigate = useNavigate();
+    let onCompleteForm = (data: {[key: string]: any}) => {
+        fetch(`/auth-complete-registration`, {
```

Comment:
I would like to see all fetches modularized so that we can swap them out. The sends should be independent of the receives, that is, you send a message and the receive happens based on an event registration. This allows the app to be fully asynchronous. Promises are quasi-async and do not allow the app to manage flow of control. You cannot ignore a promise, say, if the app state changes. You have to write special code to handle each promise.

The implementation of this type of event registration can be simple at first, e.g. based on API names.